### ISSUE TITLE
Support for "edge-case" ip notations

### ISSUE THREAD

**Issue by StoneLabs** [beaugunderson/ip-address#142]
Currently, there seem to be quite a lot of valid IPs that throw `[AddressError]: Invalid IPv4 address.`

```node
var Address4 = require('ip-address').Address4;

// All of the following are rejected as invalid.
const topic = new Address4('127.1'); // equivalent to 127.0.0.1
const topic = new Address4('127.0.1'); // equivalent to 127.0.0.1
const topic = new Address4('192.168.36095'); // equivalent to 192.168.140.255 in CLASS B notation
const topic = new Address4('192.11046143');  // equivalent to 192.168.140.255 in CLASS C notation
const topic = new Address4('0xc0.0xa8.0x8c.0xff');  // equivalent to 192.168.140.255 in valid hex notation
const topic = new Address4('0300.0250.0214.0377');  // equivalent to 192.168.140.255 in valid oct notation (though some implementations ignore oct notation)
const topic = new Address4('0000000001.0000000002.0000000003.000000010');  // equivalent to 1.2.3.8 in valid oct notation (though some implementations ignore oct notation)

// Granted, these are all very absurd notations. However, all of these are valid ip's. you can verify this just pinging them.
```

![image](https://user-images.githubusercontent.com/19885942/126387617-a17cddf1-25e2-439d-8ed4-ac9b09c9f115.png)
![image](https://user-images.githubusercontent.com/19885942/126387640-660d0090-5105-4577-a333-12fbf2b42501.png)

**Comment by beaugunderson** [comment]
I agree with you, we should successfully parse these notations… I may not
have time to add them in the short term, but welcome any PRs if someone
beats me to it!

**Comment by beaugunderson** [comment]
Closing this as a non-goal, which means retracting my 2021 comment above agreeing to it. Sorry for the reversal after leaving it open this long.

The octal half of this list is what [GHSA-mwp4-54f8-5fhr](https://github.com/beaugunderson/ip-address/security/advisories/GHSA-mwp4-54f8-5fhr) was about. `0300.0250.0214.0377` and `0000000001.0000000002.0000000003.000000010` are read as octal by the WHATWG URL parser, `inet_aton`, and `getaddrinfo`, and as decimal by `parseInt(part, 10)`, so a library that accepts them picks a host that the network stack may disagree with. 10.3.1 rejects an octet with a leading zero for that reason. Teaching the parser the rest of the notations here would reopen the same class from the other side.

The shorthand and hex forms are not a vulnerability on their own, but accepting them puts this library in the position of deciding what `127.1` and `192.11046143` name, and every reader of those forms is entitled to its own answer. SECURITY.md is explicit about the gap that leaves:

> `isValid()` returning `false` deserves particular care. It means "not a valid literal of this family". It does not mean "will not reach an internal host". `2130706433`, `0x7f000001` and `127.1` are all rejected here, and all three resolve to loopback. A guard that treats a rejected string as safe to pass along has an opening.

So the scope is narrower than what this issue asks for, deliberately: decide whether a string is a valid literal of a family and hand back an object for it, rather than emulate the notations any particular resolver accepts. Anything that needs those semantics wants the resolver itself, checking the address it is about to connect to on every hop, because a second parser with its own opinion is a source of disagreement and that disagreement is where SSRF lives.

The leading-zero case from #132 is fixed in 10.3.1. If there is a concrete use for the shorthand forms that does not involve a trust boundary, a separate issue describing it is welcome, since that would be a different question from this one.

**Comment by StoneLabs** [comment]
> Sorry for the reversal after leaving it open this long.

No worries. Thanks!
